### PR TITLE
Allows migration of DVR routers via evacuate

### PR DESCRIPTION
**Pull request by aplanas**

### CODE REVIEW

**Comment by aplanas** [comment]
@aspiers @vuntz r?

File: files/default/neutron-ha-tool.py [hunk 1]
```diff
@@ -43,14 +45,19 @@
...
     # ensure environment has necessary items to authenticate
-    for key in ['OS_TENANT_NAME', 'OS_USERNAME',
-                'OS_AUTH_URL', 'OS_REGION_NAME']:
+    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
+                'OS_ENDPOINT_TYPE']:
         if key not in os.environ.keys():
-            # We don't have a logger set up yet
-            sys.stderr.write("Your environment is missing '%s'\n" % key)
-            sys.exit(1)
+            raise SystemExit("Your environment is missing '%s'")
```
Comment (vuntz) [inline files/default/neutron-ha-tool.py]: Missing `% key`?

File: files/default/neutron-ha-tool.py [hunk 2]
```diff
@@ -43,14 +45,19 @@
...
+    '3': kclientv3
+}
+
+
 def parse_args():
     # ensure environment has necessary items to authenticate
-    for key in ['OS_TENANT_NAME', 'OS_USERNAME',
-                'OS_AUTH_URL', 'OS_REGION_NAME']:
+    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
+                'OS_ENDPOINT_TYPE']:
```
Comment (vuntz) [inline files/default/neutron-ha-tool.py]: Is it intended to lose `OS_ENDPOINT_TYPE`? Also, we need to check that `OS_PROJECT_NAME` or `OS_TENANT_NAME` is there.

File: files/default/neutron-ha-tool.py [hunk 3]
```diff
@@ -160,15 +167,38 @@ def run(args):
...
+        if not auth_version:
+            auth_version = 3
+
+    kclient = IDENTITY_API_VERSIONS[auth_version]
+    kclient_kwargs = dict()
+    kclient_kwargs['username'] = os.environ['OS_USERNAME']
+    kclient_kwargs['password'] = os.environ['OS_PASSWORD']
+    kclient_kwargs['insecure'] = args.insecure
+    kclient_kwargs['ca_cert'] = ca
+    kclient_kwargs['auth_url'] = os.environ['OS_AUTH_URL']
```
Comment (vuntz) [inline files/default/neutron-ha-tool.py:185]: We lost the region name here...

File: files/default/neutron-ha-tool.py [hunk 4]
```diff
@@ -160,15 +167,38 @@ def run(args):
...
+
+    # Instantiate Keystone client
+    keystone = kclient.Client(**kclient_kwargs)
+
+    # Instantiate Neutron client
+    qclient = nclient.Client(
+        '2.0',
+        endpoint_url=keystone.service_catalog.url_for(
+            service_type='network',
+            endpoint_type=os.environ['OS_ENDPOINT_TYPE']
```
Comment (vuntz) [inline files/default/neutron-ha-tool.py]: I'm nearly sure that our crowbar cookbook doesn't set this environment variable in the pacemaker resource (old code was using `internalURL` hard-coded, it seems).

**Review by vuntz — CHANGES_REQUESTED** [review]

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py]: Ah right. This comes from the backport. I tested it manually via command line, with the source .openrc and with the password file. Will recheck via RA.

**Comment by aplanas** [comment]
@vuntz addressed the issues, and tested in SOC7 restarting the RA

In SOC6 there is a problem in the neutron code that makes the manual and the automatic migration unreliable. I need to find the issue and backport it into the package.

File: files/default/neutron-ha-tool.py [hunk 5]
```diff
@@ -21,36 +21,47 @@
...
+    '2': kclientv2,
+    '3': kclientv3
+}
+
+
 def parse_args():
     # ensure environment has necessary items to authenticate
-    for key in ['OS_TENANT_NAME', 'OS_USERNAME',
-                'OS_AUTH_URL', 'OS_REGION_NAME']:
+    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
```
Comment (vuntz) [inline files/default/neutron-ha-tool.py]: `OS_PASSWORD` is not required as it can be passed through a file.

**Review by vuntz — CHANGES_REQUESTED** [review]
One additional fix required -- I hadn't noticed that earlier.

Everything else looks good!

File: files/default/neutron-ha-tool.py (hunk 5 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py]: I know, but is set in the RA file. Is empty but is always there.

**Review by vuntz — APPROVED** [review]

File: files/default/neutron-ha-tool.py [hunk 6]
```diff
@@ -31,8 +31,10 @@
 import sys
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:31]: [Since github does not let me review commit messages (another reason #githubsucks), I'm reviewing it here.]

Please can you update this commit message to explain where it was cherry-picked from, and also add these links:
- https://bugs.launchpad.net/openstack-ansible/+bug/1499708
- https://review.openstack.org/#/c/229053/

File: files/default/neutron-ha-tool.py (hunk 1 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: OK nice, so this is a rare valid exception to the general rule that PRs shouldn't deliberately introduce breakage and then immediately fix it.  In this case I agree it is worth cherry-picking the change from the ansible repo unchanged (except for the additions to the commit message I have suggested) and then fixing issues with the cherry-picked commit in separate commits.

File: files/default/neutron-ha-tool.py (hunk 4 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/neutron/recipes/network_agents_ha.rb#L184

File: files/default/neutron-ha-tool.py [hunk 7]
```diff
@@ -737,10 +737,21 @@ def target_agent_list(agent_list, agent_type, exclude_agent_host):
     :param exclude_agent_host: hostname of agent to exclude from the list
 
     """
+    agent_info = [
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:746]: Would be a bit clearer if this was called `excluded_agent_info`, or `source_agent_info`.

File: files/default/neutron-ha-tool.py [hunk 8]
```diff
@@ -737,10 +737,21 @@ def target_agent_list(agent_list, agent_type, exclude_agent_host):
     :param exclude_agent_host: hostname of agent to exclude from the list
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:743]: Please can you get in the habit of including full bugzilla URLs in the commit message, as per [team policy](https://etherpad.nue.suse.com/p/cloud-pull-requests)?  In this case it would be https://bugzilla.suse.com/show_bug.cgi?id=974495

File: files/default/neutron-ha-tool.py [hunk 9]
```diff
@@ -419,7 +425,7 @@ def l3_agent_migrate(qclient, noop=False, now=False):
     LOG.info("%d routers required migration from offline L3 agents",
              total_migrations)
     if total_errors > 0:
-        LOG.error("%d errors encountered during migration")
+        LOG.error("%d errors encountered during migration", total_errors)
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:428]: These were from #5 so I'm not sure why they are in this unrelated commit.  However I have now merged #5 so they will vanish on rebase.

File: files/default/neutron-ha-tool.py [hunk 10]
```diff
@@ -21,16 +21,16 @@
 #   2 - router migration required (with --l3-agent-check)
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:21]: Would be helpful to include the URL of the ansible playbook repo in the commit message.

File: files/default/neutron-ha-tool.py [hunk 11]
```diff
@@ -54,10 +54,13 @@
 
 def parse_args():
     # ensure environment has necessary items to authenticate
-    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
-                'OS_ENDPOINT_TYPE']:
+    for key in ['OS_USERNAME', 'OS_AUTH_URL', 'OS_REGION_NAME']:
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:57]: Please make the commit message explain why we're dropping `OS_ENDPOINT_TYPE`.

File: files/default/neutron-ha-tool.py [hunk 12]
```diff
@@ -171,22 +174,25 @@ def run(args):
     if not auth_version:
         auth_version = os.getenv('OS_IDENTITY_API_VERSION', None)
         if not auth_version:
-            auth_version = 3
+            auth_version = '2.0'
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:177]: I don't understand why we've cherry-picked a commit to add v3 support, and then immediately regress to v2.0?  If this is really correct then the commit message needs to explain why.

**Review by aspiers — CHANGES_REQUESTED** [review]
Thanks, the code changes basically look good but I think the commit messages could still be improved a bit as it's not 100% clear yet.

File: files/default/neutron-ha-tool.py (hunk 8 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:743]: Why the comment is in the line of code? Where do you want that I add the full link of the bug? The bug is stated in the commit message, in such a way that can be listed in the changelog.

I am not aware of this team policy etherpad, and I am not sure that OBS / IBS can track the full URL in the commit description. I would prefer to add this bsc#XXXXX signature for convenience in the commit message.

File: files/default/neutron-ha-tool.py (hunk 7 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:746]: IMHO is more clear `agent_info` here: shorter name, direct and refers to what contains and not from who. Also the variable is local, and the usage is close.

File: files/default/neutron-ha-tool.py (hunk 11 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:57]: I will.

File: files/default/neutron-ha-tool.py (hunk 12 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:177]: We add v3 from the commit, but we fallback to v2 if the version is not specified.

File: files/default/neutron-ha-tool.py (hunk 10 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:21]: Done!

File: files/default/neutron-ha-tool.py (hunk 11 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:57]: Done!

**Comment by aplanas** [comment]
@aspiers Updated the comments in the commits

File: files/default/neutron-ha-tool.py (hunk 8 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:743]: > Why the comment is in the line of code?

Because [github doesn't allow reviewing of commit messages](https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/6#discussion_r88211081).

> Where do you want that I add the full link of the bug?

In the commit message, like I said.

> The bug is stated in the commit message, in such a way that can be listed in the changelog.

Yes, but the _full URL_ is missing.

> I am not aware of this team policy etherpad

You are now, since I told you ;-)

> and I am not sure that OBS / IBS can track the full URL in the commit description.

It doesn't need to; it can track the `bsc#XXXXXX` number.

> I would prefer to add this bsc#XXXXX signature for convenience in the commit message.

I didn't ask you to drop the `bsc#XXXXXXX`.  In fact the etherpad requires that, and IMHO it's exactly right that you have already put the `bsc#XXXXXXX` in the first line of the commit message.  I was asking for the full URL to be included in the commit message body, in _addition_ to that.  The justification is given in the etherpad.

File: files/default/neutron-ha-tool.py (hunk 7 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:746]: Shorter does not imply clearer; in fact IMHO it's the exact opposite in this case: this function is dealing with multiple agents, so it's ambiguous which agent `agent_info` refers to.  The only way to understand which agent it refers to is to read the code which assigns to it.  (I'm not sure exactly what you mean by "not from who"?)

In contrast, if it was called `excluded_agent_info` then it would be self-evident, because it would obviously be referring to the docstring which says "Return a list of agents that are alive, **excluding** the one we want to migrate from".

File: files/default/neutron-ha-tool.py (hunk 10 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:21]: Thanks, but you added the URL to a different commit.  I was suggesting to add it to this commit ("neutron-ha-tool: order imports") because the commit message is:

> Ordering imports makes more easy backport patches from the
> ansible playbook repository.

so anyone reading it will immediately wonder which repo that is.

File: files/default/neutron-ha-tool.py (hunk 6 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:31]: These are the links which I was asking to be added to this backport commit ("Updated the neutron l3HA tool to use v3")

**Comment by vuntz** [comment]
@aplanas https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/5 got merged, so there's now a conflict. Can you rebase?

**Comment by aplanas** [comment]
@vuntz  rebased, thanks!

File: files/default/neutron-ha-tool.py (hunk 10 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:21]: Makes sense. Done.

File: files/default/neutron-ha-tool.py (hunk 8 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:743]: Added, but I disagree with the etherpad and the justification is weak. For one side nearly all projects refers to bugs using the local prefix associated with namespace of the project, or even with a single number. For the other side our commits from other project do not follow this rule.

> > I am not aware of this team policy etherpad
> 
> You are now, since I told you ;-)

No. Not really. When this policy was discussed? Who participate on it? When has been announced and communicated? I am not able to find this in the mailing list.

Comment (aspiers) [reply files/default/neutron-ha-tool.py:743]: > Added, but I disagree with the etherpad and the justification is weak.

Please can you explain exactly why you think it is weak?  The two reasons given in the etherpad are:
1. this makes the reference clear to outsiders who don't know what bnc / bsc is
2. it makes it possible to open the bug with a single click when viewing git history

I really struggle to imagine how you could disagree with either of these.  For the first one, remember that this is a public repository, and the code relates to a fork of some upstream code which is still in use elsewhere.  So any comments about `bsc#XXXXXX` being sufficient in internal or openSUSE-specific tools are irrelevant.  And the second one is a simple fact.

>  For one side nearly all projects refers to bugs using the local prefix associated with namespace of the project, or even with a single number. For the other side our commits from other project do not follow this rule.

Again, I am not suggesting to drop the `bsc#XXXXX` format, so how is this relevant?  I am only saying that it is worth adding the full URL in _addition_, because this brings extra clarity and convenience for anyone who reads the PR and/or the git commits.  It's not _needed_, but it is a very nice courtesy which takes very little effort and makes other people's lives easier, so why not do it?  It follows the ethos of [this great article on Empathy and Pull Requests](https://slack.engineering/on-empathy-pull-requests-979e4257d158), which I recommend reading if you didn't already.

> No. Not really.

You _are_ aware of it, even if you don't agree with it :-)

> When this policy was discussed? Who participate on it? When has been announced and communicated? I am not able to find this in the mailing list.

It was discussed a long time ago across several team workshops in Nuremberg, possibly before you joined the team, but I can't remember exactly when.

File: files/default/neutron-ha-tool.py (hunk 10 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:21]: Great, thanks!

File: files/default/neutron-ha-tool.py (hunk 11 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:57]: Thanks for the commit message update! Ahh, I misunderstood - we're not actually dropping `OS_ENDPOINT_TYPE`, but merely making it optional.  So the commit message is still slightly misleading.

File: files/default/neutron-ha-tool.py (hunk 12 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py:177]: Got it now - thanks!

File: files/default/neutron-ha-tool.py [hunk 13]
```diff
@@ -419,7 +425,7 @@ def l3_agent_migrate(qclient, noop=False, now=False):
     LOG.info("%d routers required migration from offline L3 agents",
              total_migrations)
     if total_errors > 0:
-        LOG.error("%d errors encountered during migration" % total_errors)
+        LOG.error("%d errors encountered during migration", total_errors)
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py:428]: These changes are not related to this commit.  If there is a good reason for them then I'm fine with that, but please put them in a separate commit with a message explaining the change.

**Review by aspiers — CHANGES_REQUESTED** [review]
Almost there now, thanks!  Still a couple of minor nitpicks around commit hygiene, but the main change I would like to see is adding those two URLs I suggested into the backport of the "Updated the neutron l3HA tool to use v3" commit, so that it's easier to see where it came from.

File: files/default/neutron-ha-tool.py (hunk 13 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:428]: Moved to a different commit. Done

File: files/default/neutron-ha-tool.py (hunk 6 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py:31]: Both links added. Done.

**Review by aspiers — APPROVED** [review]
Excellent!  Thanks very much for your patience with my pedantry :-)

**Comment by aspiers** [comment]
This just reminded me that we have no CI coverage for this code :-/ I guess we'll have to just trust manual testing for now...

**Comment by aplanas** [comment]
@aspiers I am very grateful for the work that you put in the reviews.

I tested it in SOC6 and SOC7 but yes, is not a trusty source. Maybe QA can fix this. @vmoravec @gosipyan @kbaikov?
